### PR TITLE
New Admin Notices [ED-692]

### PR DESCRIPTION
**Pull request by joshmarom**

### CODE REVIEW

File: core/admin/button.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Button extends \Elementor\Core\Base\Module {
+
+	private $options;
+
+	public function __construct( $options ) {
+		$this->options = $options;
```
Comment (kobizz) [inline core/admin/button.php:17]: Merge it here with the default options using `self::merge_properties()`

File: core/admin/button.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	static public function get_default_options() {
```
Comment (kobizz) [inline core/admin/button.php:20]: And then, this method shouldn't be static

File: core/admin/button.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			'icon' => '',
+			'new_tab' => false,
+			'text' => '',
+			'type' => '',
+			'url' => '',
+			'variant' => '',
+		];
+	}
+
+	private function get_options( $option = false ) {
```
Comment (kobizz) [inline core/admin/button.php:32]: And then you can just call:
`return self::get_items( $this->options, $option );`

File: core/admin/button.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		if ( isset( $this->options[ $option ] ) && array_search( $option, $this->get_default_options()  ) ) {
+			return $this->options[ $option ];
+		}
+
+		return $this->options;
+	}
+
+	public function print_button() {
+		$default_options = $this->get_default_options();
+		$options = $this->get_options();
```
Comment (kobizz) [inline core/admin/button.php:42]: Then you can rely on `$this->get_options();` only

File: core/admin/button.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+	private $options;
+
+	public function __construct( $options ) {
+		$this->options = $options;
+	}
+
+	static public function get_default_options() {
+		return [
+			'class' => 'e-button',
```
Comment (kobizz) [inline core/admin/button.php:22]: Make it an array

File: core/admin/button.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		$html .= '<span>' . sanitize_text_field( $options['text'] ) . '</span>';
+		$html .= '</' . $html_tag . '>';
+
+		echo $html;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function get_name() : string {
```
Comment (kobizz) [inline core/admin/button.php:90]: We can't use such type of hinting yet because it's only available on PHP7+

File: core/admin/button.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Button extends \Elementor\Core\Base\Module {
+
+	private $options;
+
+	public function __construct( $options ) {
```
Comment (kobizz) [inline core/admin/button.php:16]:
```suggestion
	public function __construct( array $options ) {
```

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/admin/admin.js [hunk 8]
```diff
@@ -90,15 +90,16 @@ import LandingPagesModule from 'elementor/modules/landing-pages/assets/js/admin/
 				self.animateLoader();
 			} );
 
-			$( 'div.notice.elementor-message-dismissed' ).on( 'click', 'button.notice-dismiss, .elementor-button-notice-dismiss', function( event ) {
+			$( '.e-notice--dismissible' ).on( 'click', '.e-notice__dismiss, .e-notice-dismiss', function( event ) {
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/js/admin/admin.js:93]: `e-notice__dismiss` is the class that is responsible for the x shaped dismiss button element which is a part of the notice component block.

The `e-notice-dismiss` is just an indicator class that doesn't apply any style, and can be used with other elements like a one of the buttons in the notice

File: assets/dev/scss/admin/_button.scss [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+			box-shadow: none;
+		}
+
+		&:disabled,
+		&.disabled {
+			color: $btn-link-disabled-color;
+		}
+	} */
+}
+
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_button.scss:132]: There's one instance of a button of this type in the Figma design, so I think it needs to be included. correct me if you know otherwise...

If so, I'll finish it up and commit.

File: assets/dev/scss/admin/_button.scss [hunk 10]
```diff
@@ -0,0 +1,137 @@
+@function pxToRem($size) {
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_button.scss]: Done

File: assets/dev/scss/admin/_button.scss [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+		&:disabled,
+		&.disabled {
+			color: $btn-link-disabled-color;
+		}
+	} */
+}
+
+// Future-proof disabling of clicks on `<a>` elements
+a.e-button.disabled,
+fieldset:disabled a.e-button {
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_button.scss:131]: To prevent links with button style from working when they are under the context of disabled

[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_button.scss:131]: This is the default behaviour of a real `<button>`

File: assets/dev/scss/admin/_notice.scss [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+	--e-notice-context-color: #{$editor-accent};
+	--e-notice-context-tint: var(--e-context-cta-tint-1);
+	--e-notice-box-shadow: 0 1px 4px rgba(0,0,0,.15);
+	--e-notice-dismiss-color: #{$editor-dark};
+}
+
+.e-notice {
+	position: relative;
+	display: flex;
+	font-family: Roboto, -apple-system, BlinkMacSystemFont, "Segoe UI", Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif;
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_notice.scss]: Done.

Please read about [using system-ui fonts in web apps](https://medium.com/needmore-notes/using-system-fonts-for-web-apps-bf76d214a0e0) and consider it. WordPress is doing it too...

File: assets/dev/scss/admin/_button.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_button.scss:128]: Will be saved as stash until resolution

File: core/base/db-upgrades-manager.php [hunk 13]
```diff
@@ -76,27 +77,68 @@ public function on_runner_complete( $did_tasks = false ) {
...
-		$upgrade_link = $this->get_start_action_url();
-		$message = '<p>' . sprintf( __( '%s Your site database needs to be updated to the latest version.', 'elementor' ), $this->get_updater_label() ) . '</p>';
-		$message .= '<p>' . sprintf( '<a href="%s" class="button-primary">%s</a>', $upgrade_link, __( 'Update Now', 'elementor' ) ) . '</p>';
+		/**
+		 * @var Admin_Notices $admin_notices
+		 */
+		$admin_notices = Plugin::$instance->admin->get_component( 'admin-notices' );
+
+		$options = [
+			'title' =>  $this->get_updater_label(),
```
Comment (rodolphebertozzo) [inline core/base/db-upgrades-manager.php:86]:
```suggestion
			'title' => $this->get_updater_label(),
```

**Review by kobizz — DISMISSED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
@arielk  we lose a lot of translations here. maybe it's better to leave some notices as they are?

File: assets/dev/scss/admin/_button.scss [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	}
+
+	> i {
+		line-height: inherit;
+		height: var(--e-button-line-height);
+		width: min-content;
+	}
+
+	> * + * {
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply assets/dev/scss/admin/_button.scss:98]: Add space before every child after the first one

File: core/admin/button.php [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	 * @return array|mixed
+	 */
+	private function get_options( $option = null ) {
+		return $this->get_items( $this->options, $option );
+	}
+
+	public function print_button() {
+		$options = $this->get_options();
+
+		if ( empty( $options['text'] ) ) {
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply core/admin/button.php:52]: I am not aware of this use case.
If this should be made possible, I suggest it should be done using a `hide-text` param, while still requiring the `text` param for a11y reasons

File: core/admin/admin-notices.php [hunk 16]
```diff
@@ -482,83 +472,99 @@ private function notice_role_manager_promote() {
...
+
+			<?php if ( ! empty( $options['button']['text'] ) || ! empty( $options['button_secondary']['text'] ) ) { ?>
+				<div class="e-notice__actions">
+					<?php
+					foreach ( [ $options['button'], $options['button_secondary'] ] as $index => $button_settings ) {
+						if ( empty( $button_settings['variant'] ) && $index ) {
+							$button_settings['variant'] = 'outline';
+						}
+
+						if ( empty( $button_settings['text'] ) ) {
```
[the remark this answers is not in the document]
Comment (joshmarom) [reply core/admin/admin-notices.php:555]: See [above](https://github.com/elementor/elementor/pull/13385#discussion_r554563983)

File: core/admin/button.php [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+		$html .= '</' . $html_tag . '>';
+
+		echo $html;
+	}
+
+	public function __construct( array $options ) {
+		$this->options = $this->merge_properties( $this->get_default_options(), $options );
+	}
+
+	private $options;
```
Comment (matipojo) [inline core/admin/button.php:74]: The right place for properties is before the methods.

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by kobizz — DISMISSED** [review]

**Review by matipojo — DISMISSED** [review]

**Review by kobizz — DISMISSED** [review]

**Review by rodolphebertozzo — DISMISSED** [review]

**Review by kobizz — APPROVED** [review]

File: .github/CODEOWNERS [hunk 18]
```diff
@@ -20,6 +20,7 @@
 
 # Core
 /core                             		@kingyes
+/core/admin/notices                     @nevoss
```
Comment (matipojo) [inline .github/CODEOWNERS:23]: The file is not related to the PR, can you pull `release/3.2.0`?

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — APPROVED** [review]
